File: 프로그래머스/혼자놀기의달인.java
import java.util.*;
class Solution {
    int[] parents;
    public int solution(int[] cards) {
        parents = cards;
        boolean[] visited = new boolean[cards.length];
        Map<Integer,Integer> map = new HashMap<>();
        for(int i=0;i<cards.length;i++) parents[i]--;
        
        for(int i=0;i<cards.length;i++){
            int box = 0;
            if(!visited[i]){
                box = find(i,visited);
            }else{
                box = parents[i];
            }
            map.put(box,map.getOrDefault(box,0) + 1);
        }
        List<Integer> list = new ArrayList<>();
        for(Map.Entry<Integer, Integer> entry : map.entrySet()){
            list.add(entry.getValue());
        }
        if(list.size() == 1) return 0;
        Collections.sort(list,(o1,o2) -> o2 - o1);
        
        return list.get(0) * list.get(1);
    }
    int find(int index, boolean[] visited){
        if(index == parents[index] || visited[index]){
            return index;
        }
        visited[index] = true;
        return parents[index] = find(parents[index],visited);
    }
}

File: 프로그래머스/혼자_놀기의_달인.js
function solution(cards) {
    var answer = 0;
    const visited = new Array(cards.length).fill(false);
    const groupsize = [];
    for(let i=0;i<cards.length;i++){
        let card = i;
        let size = 0;
        while(!visited[card]){
            visited[card] = true;
            card = cards[card]-1;
            size++;
        }
        groupsize.push(size);
    }
    groupsize.sort((a,b)=> b - a);
    return groupsize.length === 1 ? 0 : groupsize[0] * groupsize[1];
}
